DOC user polars instead of polars in plot_time_series_lagged_features

raisadz opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://github.com/scikit-learn/scikit-learn/blob/main/CONTRIBUTING.md
-->

#### Reference Issues/PRs
<!--
Example: Fixes #1234. See also #3456.
Please use keywords (e.g., Fixes) to create link to the issues or pull requests
you resolved, so that they will automatically be closed when your pull request
is merged. See https://github.com/blog/1506-closing-issues-via-pull-requests
-->
Related to #28341 

#### What does this implement/fix? Explain your changes.


#### Any other comments?


<!--
Please be aware that we are a loose team of volunteers so patience is
necessary; assistance handling other issues is very welcome. We value
all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.
For more information, see our FAQ on this topic:
https://scikit-learn.org/dev/faq.html#why-is-my-pull-request-not-getting-any-attention.

Thanks for contributing!
-->

raisadz commented on examples/applications/plot_time_series_lagged_features.py:
```diff
@@ -42,6 +42,22 @@
+
+import polars as pl
+
+df = pl.DataFrame({col: df[col].to_numpy() for col in df.columns})
```
I am doing this instead if `pl.from_pandas()` to avoid installing `pyarrow`.

adrinjalali commented:
Somehow I don't really love the idea of having to use both pandas and polars in the same script. It kinda defeats the purpose.

MarcoGorelli commented:
Hey,

@glemaitre  had written in the Polars Discord

>  I think this would be worth to add a pl.from_pandas in the example. This would easily accepted as visible in the documentation and in the next release for sure. Modifying fetch_openml might require more discussion regarding the type of files to read (ARFF vs. parquet), directly accessing the HTTP or caching the file, etc. So the process will be slower.

https://discord.com/channels/908022250106667068/1214872002238742529/1215228420292870164

which is why I encouraged @raisadz to try this out

glemaitre commented:
I would be fine with the conversion from pandas to polars for the moment. I think that we should make it explicit that the conversion is temporary until `fetch_openml` is natively loading `polars` frame. Once we have this feature in, then the example will perfect.

It at least shows that scikit-learn accepts such input.

It is definitely not perfect but better than not having a single example shown while awaiting for the `fetch_openml` upgrade that might not happen for the next release.

@adrinjalali WDYT?

adrinjalali commented:
@MarcoGorelli what's the status of being able to read parquet with polars w/o depending on pyarrow right now?

MarcoGorelli commented:
`pyarrow` isn't required by Polars for reading Parquet files

Here's an example of using Polars to read the greatest data science dataset in the world (no pyarrow installed!):
```python
In [1]: import polars as pl

In [2]: pl.read_parquet('iris.parquet')
Out[2]:
shape: (150, 5)
┌──────────────┬─────────────┬──────────────┬─────────────┬───────────┐
│ sepal_length ┆ sepal_width ┆ petal_length ┆ petal_width ┆ species   │
│ ---          ┆ ---         ┆ ---          ┆ ---         ┆ ---       │
│ f64          ┆ f64         ┆ f64          ┆ f64         ┆ str       │
╞══════════════╪═════════════╪══════════════╪═════════════╪═══════════╡
│ 5.1          ┆ 3.5         ┆ 1.4          ┆ 0.2         ┆ setosa    │
│ 4.9          ┆ 3.0         ┆ 1.4          ┆ 0.2         ┆ setosa    │
│ 4.7          ┆ 3.2         ┆ 1.3          ┆ 0.2         ┆ setosa    │
│ 4.6          ┆ 3.1         ┆ 1.5          ┆ 0.2         ┆ setosa    │
│ 5.0          ┆ 3.6         ┆ 1.4          ┆ 0.2         ┆ setosa    │
│ …            ┆ …           ┆ …            ┆ …           ┆ …         │
│ 6.7          ┆ 3.0         ┆ 5.2          ┆ 2.3         ┆ virginica │
│ 6.3          ┆ 2.5         ┆ 5.0          ┆ 1.9         ┆ virginica │
│ 6.5          ┆ 3.0         ┆ 5.2          ┆ 2.0         ┆ virginica │
│ 6.2          ┆ 3.4         ┆ 5.4          ┆ 2.3         ┆ virginica │
│ 5.9          ┆ 3.0         ┆ 5.1          ┆ 1.8         ┆ virginica │
└──────────────┴─────────────┴──────────────┴─────────────┴───────────┘

In [3]: import pyarrow
---------------------------------------------------------------------------
ModuleNotFoundError                       Traceback (most recent call last)
Cell In[3], line 1
----> 1 import pyarrow

ModuleNotFoundError: No module named 'pyarrow'
```

raisadz commented on examples/applications/plot_time_series_lagged_features.py:
```diff
@@ -258,36 +268,30 @@ def consolidate_scores(cv_results, scores, metric):
+            loc.body(col, extract_numeric(col) == list_min(extract_numeric(col)))
+            for col in styled_df.columns
+            if col != "loss"
+        ],
```
Replacing pandas styling with great tables as they have native support for Polars. Apart from `fetch_openml` this completely replaces pandas.

raisadz commented:
Thank you for your reviews. I added a comment about `fetch_openml` loading a pandas dataframe and it will be replaced with Polars when it will be natively supported. I replaced the pandas styling with great tables as they have Polars support and apart from `fetch_openml` this should replace pandas in this notebook.

raisadz commented:
I updated the lock files to add great tables by running `build_tools/update_environments_and_lock_files.py` and added the Polars version check to handle the deprecated syntax.

raisadz commented:
I removed checking Polars version and replaced it with using `list.get` function instead of `list.gather` (and `list.take` for earlier versions). Please, let me know if you are fine with adding `great-tables` dependency to replace the pandas styler. Otherwise, I can roll back to the previous version. Or I can just print Polars dataframes directly. Please, let me know which option you would prefer.

glemaitre commented on examples/applications/plot_time_series_lagged_features.py:
```diff
@@ -258,36 +268,33 @@ def consolidate_scores(cv_results, scores, metric):
-for col in cols_to_convert:
-    mask[col] = df[col].apply(
-        lambda x: "font-weight: bold" if x == min_values[col] else ""
+gt_styled_df = GT(styled_df)
```
For sure, we don't want the extra dependency on `great-tables`. We should instead have a way to do the same with the current installed library.

raisadz replied:
It seems that there is no way to print something in bold in Polars, that's why I added great-tables that have this support. 

Should I then convert the dataframe back to pandas when there is a styled output? As I thought you don't want to have both pandas and Polars dataframes https://github.com/scikit-learn/scikit-learn/pull/28601#issuecomment-1988098756? 

Or do you know about some other way to print that table with final metrics with bold styling?

Alternatively, should I display the dataframe without styling? I can add the minimums as a separate dataframe below.

ArturoAmorQ replied:
I guess we rather lose the bold styling than adding the extra dependency or jumping back and forth from pandas. Personally I vote for the last option then.

raisadz replied:
Thanks for your reviews. I removed the styling with great tables and added a dataframe displaying the losses that minimise each metric.

glemaitre commented on examples/applications/plot_time_series_lagged_features.py:
```diff
@@ -19,28 +19,44 @@
+# We start by loading the data from the OpenML repository
+# as a pandas dataframe. This will be replaced with Polars
+# after `fetch_openml` will add a native support for it.
+# We convert to Polars for feature engineering, as it presents several advantages
+# here:
+#
+# - It parallelises the expressions within the `with_columns` statement.
+# - It automatically caches common subexpressions which are reused in multiple
+#   expressions (like `pl.col("count").shift(1)` below). See
+#   https://docs.pola.rs/user-guide/lazy/optimizations/ for more information.
+# - It is very memory-efficient.
+# - It is stricter about data types and schemas, allowing us to catch errors earlier
+#   in our development workflow.
```
I don't think that this our place here to emphasize why one should use polars. I think that the example should only concentrate on the "consumer" part of a scikit-learn user meaning, that the underlying model will deal with a polars dataframe.

I think here the note regarding the caveats for `fetch_openml` is enough.

MarcoGorelli replied:
Hey 👋 

I just wanted to flag then when I opened https://github.com/scikit-learn/scikit-learn/pull/28576, one of the comments was

> maybe we can highlight the benefit with a comment in the code or as narrative text

So, just looping in @ArturoAmorQ to make sure that scikit-learn devs are aligned here about highlighting benefits in examples, and that you're on the same page about the direction this is going in 🤗

ArturoAmorQ replied:
I agree with @glemaitre that comments regarding the benefits should remain as minimalistic as possible to keep the focus on the example's main narrative. Maybe in this case in particular the most relevant reason is the caching of common subexpressions?

raisadz replied:
I left only the caching of common subexpressions and removed other reasons for using Polars.

Commit:
use Polars in "plot time series lagged features" example

ArturoAmorQ commented on examples/applications/plot_time_series_lagged_features.py:
```diff
@@ -258,41 +258,31 @@ def consolidate_scores(cv_results, scores, metric):
+def min_arg(col):
+    col_split = pl.col(col).str.split(" ")
+    return pl.arg_sort_by(
+        col_split.list.get(0).cast(pl.Float64),
+        col_split.list.get(2).cast(pl.Float64),
```
Honest question, why does the casting here changes the results so drastically? What is the dtype otherwise?

raisadz replied:
Thank you for your reviews @ArturoAmorQ . It looks like the fit times are slightly lower for Polars (apart from the squared loss) but otherwise there are no differences in the final table. Before casting the dtype is string (pl.Utf8). In the `extract_numeric` function that was used the mean and std were also casted to `float` before finding the minimum:

```python
def extract_numeric(value):
    parts = value.split("±")
    mean_value = float(parts[0])
    std_value = float(parts[1].split()[0])

    return mean_value, std_value
```

ArturoAmorQ commented on examples/applications/plot_time_series_lagged_features.py:
```diff
@@ -258,41 +258,31 @@ def consolidate_scores(cv_results, scores, metric):
+    ).first()
+
+
+scores_df.select(
+    pl.col("loss").get(min_arg(col_name)).alias(col_name)
+    for col_name in scores_df.columns
+    if col_name != "loss"
+)
```
This is a clever solution, though :)

ArturoAmorQ reviewed (commented):
Thanks for implementing the changes @raisadz, here's another batch of comments.

Commit:
Merge branch 'main' of github.com:scikit-learn/scikit-learn into polars_features

Commit:
change the comments about data loading and losses

ArturoAmorQ reviewed (approved):
LGTM! Any further comments on this one @glemaitre?

glemaitre reviewed (approved):
LGTM. We can further improve when we make a good OpenML fetcher using the parquet files.
